[#2070] Implement downstream address rewrite feature

dejanb opened the pull request:
Implements support for #2070

There's still documentation missing, but I'd withhold that until we have agreed upon the implementation.

sophokles73 commented on core/src/main/java/org/eclipse/hono/util/AddressHelper.java:
```diff
@@ -0,0 +1,98 @@
+ */
+public class AddressHelper {
+
+    static final String ID_REGEX = "[a-zA-Z0-9-_\\.]+";
```
is this supposed to match the regex that is configured in the registry for tenant and device IDs? If so, I believe that we will need to find a way to keep them in sync, right?

sophokles73 commented on core/src/test/java/org/eclipse/hono/util/AddressHelperTest.java:
```diff
@@ -0,0 +1,51 @@
+     * Tests address rewrite rules.
+     */
+    @Test
+    public void testAddressRewrite() {
```
maybe split this up into distinct test cases?

sophokles73 commented on core/src/main/java/org/eclipse/hono/config/ClientConfigProperties.java:
```diff
@@ -518,4 +519,24 @@ public final void setIdleTimeout(final int idleTimeoutMillis) {
+     *
+     * @param addressRewriteRule The rewrite rule to be applied to the address.
+     */
+    public void setAddressRewriteRule(final String addressRewriteRule) {
```
final ?

sophokles73 commented on client/src/main/java/org/eclipse/hono/client/impl/TelemetrySenderImpl.java:
```diff
@@ -59,32 +60,14 @@ protected TelemetrySenderImpl(
     @Override
     protected String getTo(final String deviceId) {
-        return getTargetAddress(tenantId, deviceId);
+        return AddressHelper.getTargetAddress(TelemetryConstants.TELEMETRY_ENDPOINT, tenantId, deviceId, connection.getConfig().getAddressRewriteRule());
```
how about keeping the *getTargetAddress* method and just change its implementation to delegate to `AddressHelper`. This would mean much fewer changes in the other code, wouldn't it?

dejanb replied:
@sophokles73 It would still require changes everywhere, as we need to add/pass additional argument to this method. I was having it like that in the beginning but noted that the implementation was not consistent among classes, so I moved the logic to the helper method call. At that point I didn't see any value in these intermediary methods, so I removed them.

I can reintroduce them back if you want, I don't have a strong opinion on that.

sophokles73 replied:
No worries, I see your point ...

sophokles73 commented on client/src/main/java/org/eclipse/hono/client/impl/AmqpAdapterClientFactoryImpl.java:
```diff
@@ -80,7 +83,7 @@ protected void onDisconnect() {
     public Future<TelemetrySender> getOrCreateTelemetrySender() {
 
-        final String cacheKey = TelemetrySenderImpl.getTargetAddress(tenantId, null);
+        final String cacheKey = AddressHelper.getTargetAddress(TelemetryConstants.TELEMETRY_ENDPOINT, tenantId, null, connection.getConfig().getAddressRewriteRule());
```
the `AmqpAdapterClientFactory` is supposed to be used by devices and gateways to connect to the AMQP adapter. IMHO we do not need to rewrite addresses in that case, do we?

ctron commented on core/src/main/java/org/eclipse/hono/util/AddressHelper.java:
```diff
@@ -0,0 +1,98 @@
+ */
+public class AddressHelper {
+
+    static final String ID_REGEX = "[a-zA-Z0-9-_\\.]+";
```
I think it makes sense in this case to pre-compile the regular expression using `Pattern`.

dejanb replied:
@ctron It's not a pattern at this point, but a substitute that can be used when writing rules, so you can write `telemetry/${ID}` instead of `telemetry/[a-zA-Z0-9-_\\.]+`. It's also moved now to the other place per Kai's suggestion.

dejanb commented on core/src/main/java/org/eclipse/hono/util/RegistryManagementConstants.java:
```diff
@@ -285,15 +285,20 @@
+    /**
+     * The default regular expression validating IDs contain only legal characters.
+     */
+    public static final String DEFAULT_ID_REGEX = "[a-zA-Z0-9-_\\.]+";
```
@sophokles73 it is indeed the same regex, just without end delimiters. I moved it here and based other ones on it. Let me know if you have any other ideas on how improve on this.

sophokles73 commented on core/src/main/java/org/eclipse/hono/util/AddressHelper.java:
```diff
@@ -0,0 +1,96 @@
+    static String regexSubstitues(final String endpoint, final String regex) {
+        return regex
+                .replaceAll("\\$\\{ENDPOINT\\}", endpoint)
+                .replaceAll("\\$\\{ID\\}", RegistryManagementConstants.DEFAULT_ID_REGEX);
```
what if the tenant ID and/or device ID regex are being configured explicitly, i.e. do not have their default value?

dejanb replied:
It won't work. But it's just a simple helper, so it's not something that you must use to make it work. I just thought some these would make configuration look simpler for common use-cases.

We can do more in this area and do substitutions based on configuration, environment variables, etc. But I didn't want to go there until we have the basics sorted.

If this is confusing we can remove substitutions for the time being and rethink them later on.

sophokles73 replied:
I guess it will be very hard to trace and understand the source of the problem in this case. So FMPOV we should either omit the substitution logic for the time being or integrate it properly with the configurable tenant and device ID patterns ...

dejanb replied:
OK, I'll remove it completely for now.

dejanb commented:
@sophokles73 great. Yeah, I'll do that next.

ctron commented on core/src/main/java/org/eclipse/hono/util/AddressHelper.java:
```diff
@@ -0,0 +1,82 @@
+        }
+
+        try {
+            return address.replaceAll(elements[0], elements[1]);
```
I still think we should use pre-compiled patterns for this.

dejanb replied:
It makes sense if we initialize AddressHelper and reuse that Pattern instead of using it statically. I'll take a look at that.

dejanb commented:
@sophokles73 @ctron Can you take another look?

* I moved pattern compilation to the config class, so that it will be compiled only once.
* Hence the rewrite method now have the whole config as a parameter, which I think will further help if we want to implement variable substitutions in the future. 
* I removed rewriting from a couple of places to make sure `message.setAddress()` uses the original address.
* Added docs

Let me know if you see anything else missing.

sophokles73 commented:
@dejanb The integration tests fail due to JavaDoc issues ...

sophokles73 commented on core/src/main/java/org/eclipse/hono/config/ClientConfigProperties.java:
```diff
@@ -518,4 +525,54 @@ public final void setIdleTimeout(final int idleTimeoutMillis) {
+     *
+     * @return The precompiled address rewrite pattern.
+     */
+    public Pattern getAddressRewritePattern() {
```
final?

sophokles73 commented on core/src/main/java/org/eclipse/hono/config/ClientConfigProperties.java:
```diff
@@ -518,4 +525,54 @@ public final void setIdleTimeout(final int idleTimeoutMillis) {
+     *
+     * @return The address rewrite replacement.
+     */
+    public String getAddressRewriteReplacement() {
```
final?

dejanb commented:
@sophokles73 not sure how it passed locally. Fixed now, thanks for spotting it.

sophokles73 commented:
@dejanb sorry, but thes still fail due to JavaDoc issues ...
